producer/writer/message_writer.go

```diff
@@ -312,10 +315,17 @@ func (w *messageWriterImpl) retryBatchWithLock(
 		}
 
 		w.m.writeRetry.Inc(1)
+		w.toBeRetried = append(w.toBeRetried, m)
+		retryFound++
+		if retryFound >= w.opts.MessageRetryWriteBatchSize() {
+			break
+		}
+	}
+	for _, m := range w.toBeRetried {
```

Comment:
How fast is the write here? You are writing inside the lock and I'd recommend avoiding that. Normally we simply return the slice to the caller, unlock in the caller, write outside the lock, and re-iterate, which avoids holding the lock while writing data.

Reply:
yeah, was thinking the contention will more be on the consumer writer side since the number of message writers will be bigger than number of consumer writers, but I guess I can still release lock sooner so other writes on message writer can do queue operations